File: c++/0207.cpp
// Link: https://leetcode.com/problems/course-schedule/description/
// Runtime: 16 ms
// Memory: 13.9 MB

class Solution 
{
public:
    bool canFinish(int numCourses, const std::vector<std::vector<int>>& prerequisites) 
    {
        std::vector<std::vector<int>> adjacentNodes(numCourses);
        for (const auto& connection : prerequisites)
            adjacentNodes[connection[1]].push_back(connection[0]);
        
        std::vector<bool> isVisited(numCourses), isInStack(numCourses);
        bool canFinishAllCourses = true;
        int i = 0;
        while ((i < numCourses) && (canFinishAllCourses))
        {
            if (dfs(i, adjacentNodes, isVisited, isInStack)) canFinishAllCourses = false;
            i++;
        }

        return canFinishAllCourses;
    }

private:
    bool dfs(int courseNumber, 
             const std::vector<std::vector<int>>& adjacentNodes, 
             std::vector<bool>& isVisited, 
             std::vector<bool>& isInStack)
    {
        if (isInStack[courseNumber]) return true;
        else if (isVisited[courseNumber]) return false;

        isInStack[courseNumber] = true;
        isVisited[courseNumber] = true;

        for (const auto& neighbour : adjacentNodes[courseNumber])
            if (dfs(neighbour, adjacentNodes, isVisited, isInStack))
                return true;
        
        isInStack[courseNumber] = false;

        return false;
    }
};

File: python/0207.py
# Link: https://leetcode.com/problems/course-schedule/description/
# Runtime: 116 ms
# Memory: 19.1 MB

class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adjacent_nodes = [[] for _ in range(numCourses)]
        for requirement in prerequisites:
            adjacent_nodes[requirement[1]].append(requirement[0])
        
        is_visited, is_in_stack = [False] * numCourses, [False] * numCourses
        can_finish_all_courses = True
        i = 0
        while (i < numCourses) and (can_finish_all_courses):
            if (self.__dfs(i, adjacent_nodes, is_visited, is_in_stack)):
                can_finish_all_courses = False
            i += 1
        
        return can_finish_all_courses
    

    def __dfs(self, node_index: int, adjacent_nodes: List[List[int]], is_visited: List[bool], is_in_stack: List[bool]) -> bool:
        if is_in_stack[node_index]:
            return True
        elif is_visited[node_index]:
            return False
        
        is_in_stack[node_index] = True
        is_visited[node_index] = True

        for neighbour in adjacent_nodes[node_index]:
            if self.__dfs(neighbour, adjacent_nodes, is_visited, is_in_stack):
                return True

        is_in_stack[node_index] = False

        return False
